sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4161,6 +4263,18 @@ static pi_result cleanupAfterEvent(pi_event Event) {
 
     DepEvent->WaitList.collectEventsForReleaseAndDestroyPiZeEventList(
         EventsToBeReleased);
+    if (IndirectAccessTrackingEnabled) {
+      // DepEvent has finished, we can release the associated kernel if there is
+      // one. This is the earliest place we can do this and it can't be done
+      // twice, so it is safe. Lock automatically releases when this goes out of
+      // scope.
+      std::lock_guard<std::mutex> lock(DepEvent->Queue->PiQueueMutex);
+      if (DepEvent->CommandType == PI_COMMAND_TYPE_NDRANGE_KERNEL &&
+          DepEvent->CommandData) {
+        PI_CALL(piKernelRelease(pi_cast<pi_kernel>(DepEvent->CommandData)));
```

Comment:
I don't see where you are doing `piKernelRetain` to counterpart this.

Reply:
It is retained here it piEnqueueKernelLaunch: https://github.com/intel/llvm/pull/3710/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R4001